spec/formatting.md

```diff
@@ -600,227 +608,77 @@ First, resolve the values of each _selector_:
-   « ( 1, `* bar` ), ( 0, `foo bar` ), ( 0, `foo *` ), ( 1, `* *` ) ».<br>
-   This is then sorted as:<br>
-   « ( 0, `foo bar` ), ( 0, `foo *` ), ( 1, `* bar` ), ( 1, `* *` ) ».<br>
-
-5. The pattern `Foo and bar` of the most preferred `foo bar` variant is selected.
-
-##### Selection Example 3
-
-A more-complex example is the matching found in selection APIs
-such as ICU's `PluralFormat`.
-Suppose that this API is represented here by the function `:number`.
-This `:number` function can match a given numeric value to a specific number _literal_
-and **_also_** to a plural category (`zero`, `one`, `two`, `few`, `many`, `other`)
-according to locale rules defined in CLDR.
-
-Given a variable reference `$count` whose value resolves to the number `1`
-and an `en` (English) locale,
-the pattern selection proceeds as follows for this message:
-
-```
-.input {$count :number}
-.match $count
-one {{Category match for {$count}}}
-1   {{Exact match for {$count}}}
-*   {{Other match for {$count}}}
-```
-
-1. For the selector:<br>
-   The value of the selector is resolved to an implementation-defined value
-   that is capable of performing English plural category selection on the value `1`.<br>
-   The available keys « `'one'`, `'1'` » are passed to
-   the implementation's MatchSelectorKeys method,<br>
-   resulting in a list « `'1'`, `'one'` » of matching keys.
+#### Compare Variants
 
-2. Creating the list `vars` of variants matching all keys:<br>
-   The keys of all variants are included in the list of matching keys, or use the catch-all key,<br>
-   resulting in a list « `one`, `1`, `*` » of variants.
+Next, using `res`:
 
-3. Sorting the variants:<br>
-   The list `sortable` is first set with the variants in their source order
-   and scores determined by the selector key order:<br>
-   « ( 1, `one` ), ( 0, `1` ), ( 2, `*` ) »<br>
-   This is then sorted as:<br>
-   « ( 0, `1` ), ( 1, `one` ), ( 2, `*` ) »<br>
+1. Let `bestVariant` be `UNSET`.
+1. For each _variant_ `var` of the message, in source order:
+   1. Let `keys` be the keys of `var`.
+   1. Let `match` be SelectorsMatch(res, keys).
+   1. If `match` is false:
+      1. Continue the loop.
+   1. If `bestVariant` is `UNSET`.
+      1. Set `bestVariant` to `var`.
+   1. Else:
+      1. Let `bestVariantKeys` be the keys of `bestVariant`.
+      1. If SelectorsCompare(res, keys, bestVariantKeys) is `BETTER`:
+         1. Set `bestVariant` to `var`.
+1. Select the _pattern_ of `bestVariant`.
+
+#### SelectorsMatch
+
+SelectorsMatch(selectors, keys) is defined as follows, where
+`selectors` is a list of resolved values
+and `keys` is a list of keys:
+
+1. Let `i` be 0.
+1. For each key `key` in `keys`:
+   1. If `key` is the catch-all key `'*'`
+      1. Set `i` to `i` + 1.
+      1. Continue the loop.
+   1. Let `k` be the _resolved value_ of `key` in Unicode Normalization Form C [\[UAX#15\]](https://www.unicode.org/reports/tr15).
```

Comment:
We should unwrap `k` so that the user-defined code doesn't need to do it, esp. as we know here that it's always representing a literal string.

This should also be done for the Compare() call below. It might make sense to define an abstract operation for it, rather than needing to repeat the same expression three times, together with the UAX#15 reference.

```suggestion
   1. Let `k` be the string value of the _resolved value_ of `key` in Unicode Normalization Form C [\[UAX#15\]](https://www.unicode.org/reports/tr15).
```

Reply:
That makes sense -- I added an abstract operation for it in [68311a1](https://github.com/unicode-org/message-format-wg/pull/1080/commits/68311a1b1fbf06dc5af715ea5bdf5d05b3d36c2b).